src/sentry/models/outbox.py

```diff
@@ -593,9 +593,13 @@ def process_coalesced(
         # If the context block didn't raise we mark messages as completed by deleting them.
         if coalesced is not None:
             assert first_coalesced, "first_coalesced incorrectly set for non-empty coalesce group"
-            deleted_count, _ = (
-                self.select_coalesced_messages().filter(id__lte=coalesced.id).delete()
-            )
+            deleted_count = 0
+
+            # Use a fetch and delete loop as doing cleanup in a single query
+            # causes timeouts with large datasets
+            for item in self.select_coalesced_messages().filter(id__lte=coalesced.id):
```

Comment:
how about, this:

```
# Sample batches of messages, deleting those that meet the id condition together
for items in self.select_coalesced_messages()[::100]:
  deletable = [item for item in items if item.id <= coalesced.id]
  self.filter(id__in=deletable)
```
  
```

Reply:
The core problem is that the filter by id is both *necessary* to ensure outboxes added concurrently to this processing are not deleted, but also unfortunately *linear* to length of the shard.

HOWEVER, the insight is knowing that by and large, most outboxes in the shard will either
1.  be a small total amount, so a single batch processing is not unperformant and
2.  if there are outboxes greater than the current position with respect to current processing, usually there are few of them.
3.  If there are multiple processes competing to delete at the same time, they incidentally cooperate by reducing the amount in the selected batches.

Reply:
Small performance loss in the general case, robust flattening of the worst case.

Reply:
If you just iterate, you _still_ force the database to order the items (which is linear for the entire shard size).  But you move the condition of the filter, into the python logic, the database work is reduced (incidentally total work is reduced due to concurrent cooperation that reduces the set of things to query)